### PR TITLE
Motor 417 Modernize codebase to use async/await syntax

### PR DESCRIPTION
**Pull request by prashantmital**

### CODE REVIEW

File: motor/frameworks/tornado/__init__.py [hunk 1]
```diff
@@ -48,17 +62,14 @@ def get_future(loop):
     return concurrent.Future()
 
 
-if 'MOTOR_MAX_WORKERS' in os.environ:
-    max_workers = int(os.environ['MOTOR_MAX_WORKERS'])
-else:
-    max_workers = tornado.process.cpu_count() * 5
-
-_EXECUTOR = ThreadPoolExecutor(max_workers=max_workers)
-
```
Comment (ShaneHarvey) [inline motor/frameworks/tornado/__init__.py:57]: Why change this?

File: motor/motor_gridfs.py [hunk 2]
```diff
@@ -212,15 +207,13 @@ def open(self):
            :class:`~motor.MotorGridOut` now opens itself on demand, calling
            ``open`` explicitly is rarely needed.
         """
-        return self._framework.chain_return_value(self._ensure_file(),
-                                                  self.get_io_loop(),
-                                                  self)
+        await self._ensure_file()
+        return self
```
Comment (ShaneHarvey) [inline motor/motor_gridfs.py:211]: Is this a breaking change? The docstring says "Returns a Future." but now we return an async awaitable.

File: test/asyncio_tests/__init__.py [hunk 3]
```diff
@@ -123,10 +123,9 @@ def asyncio_rsc(self, uri=None, *args, **kwargs):
             *args,
             **self.get_client_kwargs(**kwargs))
 
-    @asyncio.coroutine
-    def make_test_data(self):
-        yield from self.collection.delete_many({})
-        yield from self.collection.insert_many([{'_id': i} for i in range(200)])
```
Comment (ShaneHarvey) [inline test/asyncio_tests/__init__.py:129]: Do we need to test that `yield` still works? Perhaps we should add some tests for `yield` specifically?

File: motor/core.py [hunk 4]
```diff
@@ -1558,13 +1551,13 @@ class AgnosticLatentCommandCursor(AgnosticCommandCursor):
     __motor_class_name__ = 'MotorLatentCommandCursor'
 
     def __init__(self, collection, start, *args, **kwargs):
-        # We're being constructed without yield or await, like:
```
Comment (ShaneHarvey) [inline motor/core.py:1561]: Isn't the yield stuff still true?

File: motor/frameworks/asyncio/__init__.py [hunk 5]
```diff
@@ -48,21 +63,18 @@ def get_future(loop):
...
 
-if 'MOTOR_MAX_WORKERS' in os.environ:
-    max_workers = int(os.environ['MOTOR_MAX_WORKERS'])
-else:
-    max_workers = multiprocessing.cpu_count() * 5
-
-_EXECUTOR = ThreadPoolExecutor(max_workers=max_workers)
-
-
 def run_on_executor(loop, fn, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (prashantmital) [reply motor/frameworks/asyncio/__init__.py:66]: Done.

File: motor/frameworks/tornado/__init__.py [hunk 6]
```diff
@@ -21,12 +21,26 @@
...
 import functools
 import os
-from concurrent.futures import ThreadPoolExecutor
-
 import tornado.process
+import warnings
+
+from concurrent.futures import ThreadPoolExecutor
 from tornado import concurrent, gen, ioloop, version as tornado_version
 from tornado.gen import chain_future, coroutine  # For framework interface.
```
[the remark this answers is not in the document]
Comment (prashantmital) [reply motor/frameworks/tornado/__init__.py:29]: Motor does use `chain_future` a lot. `coroutine` is no longer used anywhere but maybe some apps import it so I have not yet removed it.

File: doc/developer-guide.rst [hunk 7]
```diff
@@ -7,49 +7,47 @@ Some explanations for those who would like to contribute to Motor development.
...
 - ``get_event_loop``
 - ``get_future``
 - ``is_event_loop``
 - ``is_future``
 - ``platform_info``
 - ``pymongo_class_wrapper``
-- ``run_on_executor``
-- ``yieldable``
+- ``run_in_executor``
+- ``run_on_executor`` (**DEPRECATED**)
```
[the remark this answers is not in the document]
Comment (prashantmital) [reply doc/developer-guide.rst:40]: Removed `run_in_executor`.

File: motor/frameworks/tornado/__init__.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply motor/frameworks/tornado/__init__.py:29]: SGTM.

File: motor/frameworks/tornado/__init__.py [hunk 8]
```diff
@@ -83,6 +71,10 @@ def chain_return_value(future, loop, return_value):
     return values resolve a Future with it, and are implemented with callbacks
     rather than a coroutine internally.
     """
+    warnings.warn(
+        "The chain_return_value function is deprecated and will be removed in"
+        "Motor 3.0", DeprecationWarning, stacklevel=2)
```
Comment (ShaneHarvey) [inline motor/frameworks/tornado/__init__.py:76]: If we're still using this method internally doesn't that mean we can't deprecate it?

File: test/asyncio_tests/__init__.py (hunk 3 above)
Comment (ShaneHarvey) [reply test/asyncio_tests/__init__.py:129]: Yeah forget this comment.

File: motor/motor_gridfs.py [hunk 9]
```diff
@@ -213,14 +210,12 @@ def open(self):
            ``open`` explicitly is rarely needed.
         """
         return self._framework.chain_return_value(self._ensure_file(),
-                                                  self.get_io_loop(),
-                                                  self)
+                                                  self.get_io_loop())
```
Comment (ShaneHarvey) [inline motor/motor_gridfs.py:213]: Missing the third argument here: `self`

Comment (prashantmital) [reply motor/motor_gridfs.py:213]: Thanks, I was just chasing down the failures myself.

File: motor/frameworks/tornado/__init__.py (hunk 8 above)
Comment (prashantmital) [reply motor/frameworks/tornado/__init__.py:76]: Correct. I removed **DEPRECATED** from the docstring and developer-guide but forgot to remove the warning.

File: motor/motor_gridfs.py (hunk 2 above)
Comment (ShaneHarvey) [reply motor/motor_gridfs.py]: Can you open a ticket for this change? Perhaps we'll think of a better solution later.

**Pull request merged**